### PR TITLE
crimson/osd/pg: do_delete_work use claim_and_reset 

### PR DESCRIPTION
**Pull request by Matan-B**
Fixes ceph_test_cls_rbd (cyanstore) crash.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/crimson/osd/pg.cc [hunk 1]
```diff
@@ -433,10 +433,11 @@ PG::do_delete_work(ceph::os::Transaction &t, ghobject_t _next)
...
     t.remove(coll_ref->get_cid(), pgmeta_oid);
     t.remove_collection(coll_ref->get_cid());
-    (void) shard_services.get_store().do_transaction(
-      coll_ref, std::move(t)).then([this] {
-      return shard_services.remove_pg(pgid);
-    });
+    t.register_on_commit(
+      new LambdaContext([this](int r) {
+      ceph_assert(r == 0);
+        (void) shard_services.remove_pg(pgid);
```
Comment (athanatos) [inline src/crimson/osd/pg.cc:439]: Spacing looks odd.

**Comment by athanatos** [comment]
Are you sure the problem isn't that the continuation isn't capturing a refcount to the pg?

**Comment by Matan-B** [comment]
> Are you sure the problem isn't that the continuation isn't capturing a refcount to the pg?

@athanatos, It doesn't seem like pg lifetime issue. It looks like the transaction is being used again after being moved (added test object):
```
diff --git a/src/crimson/osd/pg.cc b/src/crimson/osd/pg.cc
@@ -432,6 +432,8 @@ PG::do_delete_work(ceph::os::Transaction &t, ghobject_t _next)
   if (objs_to_rm.empty()) {
     logger().info("all objs removed, removing coll for {}", pgid);
     t.remove(coll_ref->get_cid(), pgmeta_oid);
+    ghobject_t tst(hobject_t(sobject_t("TST", CEPH_NOSNAP)));
+    t.touch(coll_ref->get_cid(), tst);
     t.remove_collection(coll_ref->get_cid());
     (void) shard_services.get_store().do_transaction(
       coll_ref, std::move(t)).then([this] {
```

```
$ cat osd.2.log|grep -E 'cyanstore|dispatch_context_transaction|removed' | tail -n 50
DEBUG 2024-01-25 14:29:23,941 [shard 1] cyanstore - list_objects 2.11_head GHMAX GHMAX 30
INFO  2024-01-25 14:29:23,941 [shard 1] osd - all objs removed, removing coll for 2.11
DEBUG 2024-01-25 14:29:23,941 [shard 1] cyanstore - _remove cid=2.11_head oid=#2:88000000::::head#
DEBUG 2024-01-25 14:29:23,941 [shard 1] cyanstore - _get_collection cid=2.11_head
DEBUG 2024-01-25 14:29:23,941 [shard 1] cyanstore - _touch cid=2.11_head oid=#-1:99a3f4e0:::TST:head#           <-----
DEBUG 2024-01-25 14:29:23,941 [shard 1] cyanstore - _get_collection cid=2.11_head
DEBUG 2024-01-25 14:29:23,941 [shard 1] cyanstore - _remove_collection cid=2.11_head
DEBUG 2024-01-25 14:29:23,941 [shard 1] cyanstore - _get_collection cid=2.11_head

DEBUG 2024-01-25 14:29:23,941 [shard 1] osd - ShardServices::dispatch_context_transaction: do_transaction ...
DEBUG 2024-01-25 14:29:23,941 [shard 1] cyanstore - _remove cid=2.11_head oid=#2:88000000::::head#
DEBUG 2024-01-25 14:29:23,941 [shard 1] cyanstore - _get_collection cid=2.11_head

DEBUG 2024-01-25 14:29:23,941 [shard 1] cyanstore - _touch cid=2.11_head oid=#-1:99a3f4e0:::TST:head#         <-----
DEBUG 2024-01-25 14:29:23,941 [shard 1] cyanstore - _get_collection cid=2.11_head
ERROR 2024-01-25 14:29:23,941 [shard 1] cyanstore -  transaction dump:
```

First do_transaction caller is from `PG::do_delete_work` and the second is from `ShardServices::dispatch_context_transaction`.

IIUC, registering on commit will avoid the double usage.

**Comment by athanatos** [comment]
Ah, I see.  You also need to capture a reference to the pg, regardless.

**Review by athanatos — APPROVED** [review]
LGTM, capture a reference to this.

**Comment by Matan-B** [comment]
> LGTM, capture a reference to this.

@athanatos, can you clarify please?

**Review by athanatos — APPROVED** [review]

**Review by xxhdx1985126 — APPROVED** [review]

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by Matan-B** [comment]
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-do_delete_work

File: src/crimson/osd/pg.cc [hunk 2]
```diff
@@ -434,7 +434,7 @@ PG::do_delete_work(ceph::os::Transaction &t, ghobject_t _next)
     t.remove(coll_ref->get_cid(), pgmeta_oid);
     t.remove_collection(coll_ref->get_cid());
     (void) shard_services.get_store().do_transaction(
-      coll_ref, std::move(t)).then([this] {
```
Comment (cyx1231st) [inline src/crimson/osd/pg.cc:437]: I'm not following the difference between moving `t` and calling `t.claim_and_reset()` (which looks to be a swap() internally).

Appriciated if you can help elaborate this.

Comment (Matan-B) [reply src/crimson/osd/pg.cc:437]: See https://github.com/ceph/ceph/pull/55305#issuecomment-1910350264
we move `t` and then use it again without resetting. `claim_and_reset` will move **and** reset the txn.

Comment (cyx1231st) [reply src/crimson/osd/pg.cc:437]: Oh, I think you mean that `ShardServices::dispatch_context_transaction` will reuse the same transaction data-structure after `PG::do_delete_work`. UB will happen if it is not reset. Thanks!

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2024-01-30_11:35:05-crimson-rados-wip-matanb-crimson-do_delete_work-distro-crimson-smithi/

**Pull request merged**